### ISSUE TITLE
Bug: Impossible to configure flat config for folder overrides?

### ISSUE THREAD

**Issue by TotallyInformation** [eslint/eslint#19456]
### Environment

Node version: v20.18.1
npm version: N/A
Local ESLint version: 9.20.1
Global ESLint version: None
Operating System: Windows 11

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
export default [
    {
        languageOptions: {
            // sourceType: 'script',
            ecmaVersion: 2018,
        },
    },
    {
        files: ['**/*'],
        languageOptions: {
            sourceType: 'script',
            ecmaVersion: 'latest',
        },
    },
    {
        files: ['nodes/**/*'],
        languageOptions: {
            sourceType: 'commonjs',
            ecmaVersion: 2019,
        },
    },
]
```
</details>

I have tried the above configurations in multiple orders and combinations.

### What did you expect to happen?

Any js file in the `nodes/` folder or any of its sub-folders should be treated as commonjs using ES2019. 

Files in other folders or in the root should be treated as `module` (which seems to be the current default) and using the `latest` ES version.

### What actually happened?

Any js file in the `nodes/` folder or any of its sub-folders is treated according to the default or the section with  `files: ['**/*']`.

The section with `files: ['nodes/**/*']` is completely ignored.

`files: ['nodes/**/*']` can be written as `files: ['nodes']`  or `files: ['nodes/*']`  and it is still completely ignored no matter what order the configs come in.

### Link to Minimal Reproducible Example

https://github.com/TotallyInformation/node-red-contrib-uibuilder

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by TotallyInformation** [comment]
I've also tried adding `ignores: ['nodes/**/*.js'],` to one or both of the non-relevant configs and that doesn't work either.

**Comment by mdjermanovic** [comment]
Hi @TotallyInformation, thanks for the issue!

> https://github.com/TotallyInformation/node-red-contrib-uibuilder

There doesn't seem to be an `eslint.config.js` file there. Can you please provide a branch where this problem is observable?

**Comment by TotallyInformation** [comment]
Sorry, I don't think I pushed my v9 config because it doesn't work. I will try to create a simple example repo in the morning.

**Comment by mdjermanovic** [comment]
Can you clarify for which files in the repro the config isn't as expected? 

`node-mjs\node-mjs.mjs` from the screenshot doesn't match any of the `files` patterns in `eslint.config.mjs`, and the file doesn't exist.

**Comment by mdjermanovic** [comment]
Note that it does match the config object that doesn't have `files`, so the resulting `sourceType` is `'script'` and `ecmaVersion` is `2018`.

https://github.com/TotallyInformation/eslintv9test/blob/af16674bf626a253064c6f0d1fcb5ff333992438/eslint.config.mjs#L24-L30

**Comment by TotallyInformation** [comment]
My bad, rushing this morning.

Corrected config:

```js
export default [
  {
    // default (browser-script): browser/script/2018
    // ignores: ['browser-module/**/*.js'],
    languageOptions: {
      sourceType: 'script',
      ecmaVersion: 2018,
    },
  },
  {
    // browser-module: browser/module/2019
    files: ['browser-module/**/*'],
    languageOptions: {
      sourceType: 'module',
      ecmaVersion: 2019,
    },
  },
  {
    // node-cjs: node/commonjs/2020
    files: ['node-cjs/**/*'],
    languageOptions: {
      sourceType: 'commonjs',
      ecmaVersion: 2019,
    },
  },
  {
    // node-esm: node/esm/latest
    files: ['node-esm/**/*'],
    languageOptions: {
      sourceType: 'module',
      ecmaVersion: 'latest',
    },
  },
];
```

Also pushed to repo.

`node-esm\node-esm.mjs` should match the last config `files: ['node-esm/**/*'],` - but it does not

![Image](https://github.com/user-attachments/assets/0e0371b8-439a-458b-ad1b-d6a01dcd2274)

**Comment by mdjermanovic** [comment]
In ESLint runtime, `node-esm\node-esm.mjs` does match the config with `files: ['node-esm/**/*']`, and `sourceType: 'module'` and `ecmaVersion: 'latest'` are applied:

```cmd
C:\projects\tmp\eslintv9test>npx eslint --print-config node-esm\node-esm.mjs
{
  "linterOptions": {
    "reportUnusedDisableDirectives": 1
  },
  "rules": {},
  "plugins": [
    "@"
  ],
  "language": "@/js",
  "languageOptions": {
    "sourceType": "module",
    "ecmaVersion": 2025,
    "parser": "espree@10.3.0",
    "parserOptions": {
      "sourceType": "module"
    }
  }
}
```

This seems to be a problem in ESLint Config Inspector, so I'm transferring this issue to its repo.
